### ISSUE TITLE
wav2vec model VALID_WER cannot go down after reaching 100 when fine-tuneing with own dataset

### ISSUE THREAD

**Issue by MorrisXu-Driving** [pytorch/fairseq#2570]
#### What is your question?
I am trying to fine-tune the wav2vec_small.pt model based on my own dataset. However, as the training goes on, the wer cannot go down further as it reach 100.
**Dict:**
![image](https://user-images.githubusercontent.com/52644306/92210441-4c095400-eec1-11ea-9e51-1246bd9c0cba.png)

**Label**
![image](https://user-images.githubusercontent.com/52644306/92210416-40b62880-eec1-11ea-9dca-3ad3ba24c315.png)

**Manifest**
![image](https://user-images.githubusercontent.com/52644306/92210482-5b889d00-eec1-11ea-984e-9e4b5e93ad79.png)

**Training Log:**
![捕获](https://user-images.githubusercontent.com/52644306/92210080-ab1a9900-eec0-11ea-9dbb-93f62b4e729f.JPG)

Could anyone tell me what could be the potential mistake and how to solve it?

**Comment by lematt1991** [comment]
Can you please provide a reproducible example including the full command that you ran

**Comment by alexeib** [comment]
can you take one of the checkpoints and decode it to see what the model is predicting? your UER is also 100 which means it can't predict any letters at all

**Comment by luweishuang** [comment]
I encounted the same question and the "pred_units_arr" in fairseq/criterions/ctc.py kept  the null during training. Here is my training log 
[nohup.zip](https://github.com/pytorch/fairseq/files/5192891/nohup.zip)

**Comment by alexeib** [comment]
> I encounted the same question and the "pred_units_arr" in fairseq/criterions/ctc.py kept the null during training. Here is my training log

I looked at the log and i see you are printing out pred_units_arr. this is only done during validation to make training reasonably quick (that is why its not populated during training)

i guess you are working with mandarin, and have a much larger vocab (8583 units?). I am not sure how much data you are fine-tuning on, but if it some small amount it may not work with a vocab this big. You might want to use something like phones instead and then decode to words later. You can also try the seq2seq arch (wav2vec_seq2seq instead of wav2vec_ctc) but you will have to tune hyper params (and use a bigger fine-tuning dataset)

**Comment by luweishuang** [comment]
@alexeib  Thanks for replying. I do trained a mandarin asr model and first I used mandarin char as units(8583). As you suggested, I used pinyin as recognition unit(dict.ltr.txt keep the same as english letter file) instead of mandarin char and I used more large training dataset, which have 200 hours instead of 10 hours before. Before I used to training wav2vec_ctc in this 200 hours dataset and using wav2vec_vox.pt as w2v_model and it got a little less result than wav2letter. But when I use the same 200 hours training dataset to train wav2vec_ctc with w2v_small.pt model, It got a significant worse asr result. In fact I never got training done for dev_uer always gone out of control. Here is my training command `
python  train.py --distributed-world-size 2 examples/wav2vec/output/all20_py --save-dir examples/wav2vec/models/mandarin/ctc/ctc9 --post-process letter --valid-subset dev  --best-checkpoint-metric wer --num-workers 8 --max-update 320000 --sentence-avg --task audio_pretraining --arch wav2vec_ctc --w2v-path examples/wav2vec/models/mandarin/wav2vec_small.pt --labels ltr --apply-mask --mask-selection static --mask-other 0 --mask-length 10 --mask-prob 0.5 --layerdrop 0.1 --mask-channel-selection static --mask-channel-other 0 --mask-channel-length 64 --mask-channel-prob 0.5 --zero-infinity --feature-grad-mult 0.0 --freeze-finetune-updates 10000 --validate-after-updates 10000 --optimizer adam --adam-betas "(0.9, 0.98)" --adam-eps 1e-08 --lr 2e-05 --lr-scheduler tri_stage --warmup-steps 32000 --hold-steps 32000 --decay-steps 40000 --final-lr-scale 0.05 --final-dropout 0.0 --dropout 0.0 --activation-dropout 0.1 --criterion ctc --attention-dropout 0.0 --max-tokens 3840000 --seed 2337 --log-format json --log-interval 6000 --ddp-backend no_c10d`

**Comment by haythemdhieb** [comment]
Any solution. I have the same problem. The training stops when   WER  reaches 100.

**Comment by AdrianVandierAst** [comment]
Hello :)
We face the same problem but with librispeech dataset. We use the default configs (base_100h or base_960h).
When using base_960h with complete librispeech dataset everything is fine, the raw_wer is decreasing slowly to around 8. But when using base_100h with only test-clean-100, raw_wer is always above 100. We tried to modify the config in a lot of ways and we never succeed to get below 100.

What can be the reason ? Any clue on that 100 threshold ?

**Comment by maggieezzat** [comment]
@alexeib excuse me but what does UER exactly stand for?

**Comment by alexeib** [comment]
unit error rate - the error rate for whatever units the model is predicting (letters, BPE tokens, etc whatever) as opposed to "word error rate" where these units are composed into words before evaluating the error rate. usually the unit error rate is significantly lower, and it is also more appropriate for languages where space tokenization is problematic (e.g. many asian languages)

**Comment by maggieezzat** [comment]
but what does UER above 100 mean? It has also been increasing actually.
I am getting weird results for the validation UER (training on 25 hr and validation on 4 hr):

`
2021-01-25 11:53:19 | INFO | valid | {"epoch": 4, "valid_loss": "898.216", "valid_ntokens": "9023.83", "valid_nsentences": "309.083", "valid_nll_loss": "30.766", "valid_uer": "200.395", "valid_wer": "100", "valid_raw_wer": "100", "valid_wps": "29902.6", "valid_wpb": "9023.8", "valid_bsz": "309.1", "valid_num_updates": "50"}

2021-01-25 11:54:55 | INFO | valid | {"epoch": 7, "valid_loss": "893.164", "valid_ntokens": "9023.83", "valid_nsentences": "309.083", "valid_nll_loss": "30.593", "valid_uer": "206.089", "valid_wer": "100", "valid_raw_wer": "100", "valid_wps": "22975.2", "valid_wpb": "9023.8", "valid_bsz": "309.1", "valid_num_updates": "100", "valid_best_wer": "100"}

2021-01-25 11:56:30 | INFO | valid | {"epoch": 10, "valid_loss": "885.232", "valid_ntokens": "9023.83", "valid_nsentences": "309.083", "valid_nll_loss": "30.321", "valid_uer": "218.975", "valid_wer": "100", "valid_raw_wer": "100", "valid_wps": "28383.4", "valid_wpb": "9023.8", "valid_bsz": "309.1", "valid_num_updates": "150", "valid_best_wer": "100"}

2021-01-25 11:58:05 | INFO | valid | {"epoch": 13, "valid_loss": "873.498", "valid_ntokens": "9023.83", "valid_nsentences": "309.083", "valid_nll_loss": "29.919", "valid_uer": "232.425", "valid_wer": "100", "valid_raw_wer": "100", "valid_wps": "25443.6", "valid_wpb": "9023.8", "valid_bsz": "309.1", "valid_num_updates": "200", "valid_best_wer": "100"}

2021-01-25 11:59:38 | INFO | valid | {"epoch": 16, "valid_loss": "857.41", "valid_ntokens": "9023.83", "valid_nsentences": "309.083", "valid_nll_loss": "29.368", "valid_uer": "228.627", "valid_wer": "100", "valid_raw_wer": "100", "valid_wps": "28099.1", "valid_wpb": "9023.8", "valid_bsz": "309.1", "valid_num_updates": "250", "valid_best_wer": "100"}

`

**Comment by maggieezzat** [comment]
I am facing the same problem unfortunately. Using the XLSR pretrained model to finetune on arabic dataset. The UER as well as the WER is 100

**Comment by alexeib** [comment]
try setting freeze_finetuning_updates to 0 so you can see progress quicker.

**Comment by muntasir2000** [comment]
I am also facing the same issue when fine tuning the XLSR-53 model on Bengali Babel data (82hr). I'm using base_100h config. But the WER and and UER gets stuck at 100 after epoch 2. But the loss is decreasing.

My config and log  - 
[hydra_train.log](https://github.com/pytorch/fairseq/files/5936598/hydra_train.log)

I have also tried using other parameter configurations (max_tokens, mask_channel_prob, freeze_finetune_updates to zero) etc. Any help appreciated.

**Comment by zxpan** [comment]
Anyone have any insights how to proceed with the issue (valid wer stays 100)?

**Comment by zxpan** [comment]
seem dict.ltr.txt matters here, how the order (letter) of symbol plays the role here?  @alexeib @lematt1991

**Comment by alexeib** [comment]
try printing the output of the checkpoint and see what it is producing. unfortunately it is difficult to help since i dont know your setup, you will have to debug :(

**Comment by zxpan** [comment]
just FYI: with latest fairseq code, I am now seeing the training in right direction (loss decreasing)

**Comment by vigneshgig** [comment]
Hi, I am also getting a valid WER of 100. And also I checked the predicted dataset, I am getting a single character. 
BOXER ENGINE	O
BOOST	O
BLUETOOTH	O
BLIND SPOT	I
BETWEEN	I
BEFORE NOON	I
BEFORE LUNCH	R
BED LINER	E
BASE MODEL	I
BASE	O
BALENO ZETA CVT	R
BALENO ZETA	O
BALENO SIGMA	O
BALENO DELTA CVT	R
BALENO DELTA	E
BALENO ALPHA CVT	E
BALENO ALPHA	O
BACKUP CAMERA	O
AWB	O
@alexeib
[log.txt](https://github.com/pytorch/fairseq/files/6135229/log.txt)
Thanks

**Comment by vigneshgig** [comment]
> Hi, I am also getting a valid WER of 100. And also I checked the predicted data, below as you see I am getting a single character.
> BOXER ENGINE O
> BOOST O
> BLUETOOTH O
> BLIND SPOT I
> BETWEEN I
> BEFORE NOON I
> BEFORE LUNCH R
> BED LINER E
> BASE MODEL I
> BASE O
> BALENO ZETA CVT R
> BALENO ZETA O
> BALENO SIGMA O
> BALENO DELTA CVT R
> BALENO DELTA E
> BALENO ALPHA CVT E
> BALENO ALPHA O
> BACKUP CAMERA O
> AWB O
> @alexeib
> [log.txt](https://github.com/pytorch/fairseq/files/6135229/log.txt)
> Thanks

Hi , I installed latest fairseq and pytorch 1.7.0 with cuda 10.2 in python 3.7.9 version.
Now the valid_wer is decreasing and also it starts from 83~ .
and also trained loss start from 54.0 comparing previous older version 117~
[Uploading hydra_train.log…]()

**Comment by vigneshgig** [comment]
> > Hi, I am also getting a valid WER of 100. And also I checked the predicted data, below as you see I am getting a single character.
> > BOXER ENGINE O
> > BOOST O
> > BLUETOOTH O
> > BLIND SPOT I
> > BETWEEN I
> > BEFORE NOON I
> > BEFORE LUNCH R
> > BED LINER E
> > BASE MODEL I
> > BASE O
> > BALENO ZETA CVT R
> > BALENO ZETA O
> > BALENO SIGMA O
> > BALENO DELTA CVT R
> > BALENO DELTA E
> > BALENO ALPHA CVT E
> > BALENO ALPHA O
> > BACKUP CAMERA O
> > AWB O
> > @alexeib
> > [log.txt](https://github.com/pytorch/fairseq/files/6135229/log.txt)
> > Thanks
> 
> Hi , I installed latest fairseq and pytorch 1.7.0 with cuda 10.2 in python 3.7.9 version.
> Now the valid_wer is decreasing and also it starts from 83~ .
> and also trained loss start from 54.0 comparing older version which starts from 117~
> Uploading hydra_train.log…

This wer problem is not problem of version. It's problem of freeze_finetuning_updates parameter. please read this
https://github.com/pytorch/fairseq/issues/2685#issuecomment-702920714
